WalletWasabi/Crypto/ZeroKnowledge/Prover.cs

```diff
@@ -11,40 +11,65 @@ namespace WalletWasabi.Crypto.ZeroKnowledge
 {
 	public static class Prover
 	{
+		public const string KnowledgeOfRepresentationTag = "KnowledgeOfRepresentation"; // TODO add precise statement expression
+
 		public static KnowledgeOfDiscreteLog CreateProof(Scalar secret, GroupElement publicPoint, GroupElement generator, WasabiRandom? random = null)
 		{
-			var proof = CreateProof(new[] { (secret, generator) }, publicPoint, random);
+			return CreateProof(new Transcript(), secret, publicPoint, generator, random);
+		}
+
+		public static KnowledgeOfDiscreteLog CreateProof(Transcript transcript, Scalar secret, GroupElement publicPoint, GroupElement generator, WasabiRandom? random = null)
+		{
+			var proof = CreateProof(transcript, new[] { (secret, generator) }, publicPoint, random);
 
 			return new KnowledgeOfDiscreteLog(proof.Nonce, proof.Responses.First());
 		}
 
 		public static KnowledgeOfRepresentation CreateProof(IEnumerable<(Scalar secret, GroupElement generator)> secretGeneratorPairs, GroupElement publicPoint, WasabiRandom? random = null)
 		{
-			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+			return CreateProof(new Transcript(), secretGeneratorPairs, publicPoint, random);
+		}
 
+		public static KnowledgeOfRepresentation CreateProof(Transcript transcript, IEnumerable<(Scalar secret, GroupElement generator)> secretGeneratorPairs, GroupElement publicPoint, WasabiRandom? random = null)
+		{
+			var generators = secretGeneratorPairs.Select(x => x.generator);
 			var nonce = GroupElement.Infinity;
 			var randomScalars = new List<Scalar>();
 			var publicPointSanity = publicPoint;
+
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+
 			foreach (var (secret, generator) in secretGeneratorPairs)
 			{
 				Guard.False($"{nameof(secret)}.{nameof(secret.IsOverflow)}", secret.IsOverflow);
 				Guard.False($"{nameof(secret)}.{nameof(secret.IsZero)}", secret.IsZero);
 				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
 				publicPointSanity -= secret * generator;
-
-				var randomScalar = GetNonZeroRandomScalar(random);
-				randomScalars.Add(randomScalar);
-				var randomPoint = randomScalar * generator;
-				nonce += randomPoint;
 			}
 
 			if (publicPointSanity != GroupElement.Infinity)
 			{
 				throw new InvalidOperationException($"{nameof(publicPoint)} was incorrectly constructed.");
 			}
 
-			var generators = secretGeneratorPairs.Select(x => x.generator);
-			var challenge = Challenge.Build(publicPoint, nonce, generators);
+			// before modifying anything, save a copy of the initial transcript state
+			// for the verification check below
+			var transcriptCopy = transcript.Clone();
+
+			transcript.Statement(Encoding.UTF8.GetBytes(KnowledgeOfRepresentationTag), publicPoint, generators);
```

Comment:
With this we are changing the internal state of the transcript but there are late validations that can throw exceptions and break the flow of execution. Should we restore the transcript to its previous state in case of error?

Reply:
i think there's no need because:

1. this will get refactored before then, in fact in my working copy i think all the guards are already before everything already
2. generally if one proof is invalid the entire request will fail anyway